src/swish/parallel.ss

```diff
@@ -0,0 +1,378 @@
+;;; included in all copies or substantial portions of the Software.
+;;;
+;;; THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+;;; EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+;;; MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+;;; NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+;;; HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+;;; WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+;;; OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+;;; DEALINGS IN THE SOFTWARE.
+
+(library (swish parallel)
+  (export
+   parallel
+   parallel!
+   parallel:execute
+   parallel:execute!
+   parallel:for-effect
+   parallel:map
+   parallel:options
+   parallel:vector-map
+   )
+  (import
+   (chezscheme)
+   (swish erlang)
+   (swish event-mgr-notify)
+   (swish events)
+   (swish options)
+   )
+
+  (define-options parallel:options
+    (optional
+     [start-limit
+      (default (most-positive-fixnum))
+      (must-be fixnum? fxpositive?)]
+     [order
+      (default 'random)
+      (must-be symbol? (lambda (x) (memq x '(random left right))))]
+     [timeout
+      (default 'infinity)
+      [must-be
+       (lambda (x)
+         (or (eq? x 'infinity)
+             (and (fixnum? x) (fxnonnegative? x))))]]))
+
+  (define default-options (parallel:options))
+
+  (define (shuffle-vector! v)
+    (let lp ([i (vector-length v)])
+      (when (fx> i 1)
+        (let* ([j (random i)]
+               [i (fx- i 1)]
+               [t (vector-ref v i)])
+          (vector-set! v i (vector-ref v j))
+          (vector-set! v j t)
+          (lp i))))
+    v)
+
+  (define (build-index-vector len order)
+    (let ([vec (make-vector len)])
+      (match order
+        [random
+         (do ([i 0 (fx+ i 1)]) ((fx= i len) vec)
+           (vector-set! vec i i))
+         (shuffle-vector! vec)]
+        [left
+         (do ([i 0 (fx+ i 1)]) ((fx= i len) vec)
+           (vector-set! vec i i))]
+        [right
+         (do ([i 0 (fx+ i 1)]) ((fx= i len) vec)
+           (vector-set! vec i (fx- len i 1)))])))
+
+  (define ($scatter/gather! opt f vec)
+    (match-define `(<parallel:options> ,start-limit ,order ,timeout) opt)
+    (define me self)
+    (define len (vector-length vec))
+    (define index-vec (build-index-vector len order))
+    (define monitors (make-eq-hashtable))
+    (define expire-time
+      (if (eq? timeout 'infinity)
+          'infinity
+          (+ (erlang:now) timeout)))
+    (define parent-pid (process-parent))
+    (define (wait-for-any expire-time raise?)
+      (receive (until expire-time (throw 'timeout))
+        [`(DOWN ,m ,_ ,reason ,err)
+         (cond
+          [(eq-hashtable-ref monitors m #f)
+           (eq-hashtable-delete! monitors m)
+           (when (and raise? (not (eq? reason 'normal)))
+             (raise err))]
+          [else
+           (wait-for-any expire-time raise?)])]
```

Comment:
Do you need the monitors since you'll get the EXIT messages?

Reply:
```
  (define ($scatter/gather! opt f vec)
    (match-define `(<parallel:options> ,start-limit ,order ,timeout) opt)
    (define me self)
    (define len (vector-length vec))
    (define index-vec (build-index-vector len order))
    (define pid-ht (make-eq-hashtable))
    (define expire-time
      (if (eq? timeout 'infinity)
          'infinity
          (+ (erlang:now) timeout)))
    (define parent-pid (process-parent))
    (define (wait-for-any expire-time raise?)
      (receive (until expire-time (throw 'timeout))
        [`(EXIT ,pid ,reason ,err)
         (cond
          [(and raise? (eq? pid parent-pid))
           ;; When the kernel is running and the parent dies, escape
           ;; so the kernel can stop the workers and die for the
           ;; same reason.
           (raise err)]
          [(eq-hashtable-ref pid-ht pid #f)
           (eq-hashtable-delete! pid-ht pid)
           (when (and raise? (not (eq? reason 'normal)))
             (raise err))]
          [else
           (wait-for-any expire-time raise?)])]))
    (define (start-work idx)
      (let* ([i (vector-ref index-vec idx)]
             [v (vector-ref vec i)]
             [pid (spawn&link (lambda () (vector-set! vec i (f v))))])
        (eq-hashtable-set! pid-ht pid 0)))
    (define (sg started)
      (let ([running (hashtable-size pid-ht)])
        (cond
         [(and (fx= started len) (= running 0))
          vec]
         [(and (fx< started len) (< running start-limit))
          (start-work started)
          (sg (fx+ started 1))]
         [else
          (wait-for-any expire-time #t)
          (sg started)])))
    (match (try (sg 0))
      [`(catch ,_ ,err)
       (let ([pids (hashtable-keys pid-ht)])
         (when (> (vector-length pids) 0)
           (vector-for-each (lambda (pid) (kill pid err)) pids)
           ;; Wait for all workers to complete in case they are
           ;; holding critical resources needed by any subsequent
           ;; user-code.
           (vector-for-each (lambda (pid) (wait-for-any 'infinity #f)) pids)))
       (raise err)]
      [,result result]))
```

Reply:
Thanks for reminding me of why supervisors use monitors in addition to links. I very rarely use unlink, so it seems like a reasonable caveat. I wonder what the memory & time savings would be for eliminating the monitors? I suspect it'll be very small.

Reply:
I'd say keep the monitors so that the supervision is consistent with the supervisor module.

Reply:
I didn't read the supervisor code carefully enough. It doesn't use monitors until shutdown time.

Reply:
We'll go with the "no monitor" solution. The reduction in code complexity won. There is some mild performance improvement. We currently don't have a lot of "real world" examples to measure.

The very simple, but unlikely `(parallel () (unlink (process-parent)))` will cause `parallel` to never return. Since supervisors and event-mgr are also susceptible to naughty children unlinking, we may want to address that in a future release. It would be nice to have a design pattern we can rely on without storing tables of monitors.